hudi-utilities/src/test/java/org/apache/hudi/utilities/deltastreamer/TestHoodieDeltaStreamer.java

```diff
@@ -670,6 +676,265 @@ public void testSchemaEvolution(String tableType, boolean useUserProvidedSchema,
+      cfg.sourceLimit = 100_000;
+      cfg.ignoreCheckpoint = "12345";
+      new HoodieDeltaStreamer(cfg, jsc).sync();
+      logicalAssertions(tableSchema, tableBasePath, hudiOpts, version.versionCode());
+    });
+  }
+
+  private void logicalAssertions(Schema tableSchema, String tableBasePath, Map<String, String> hudiOpts, int tableVersion) {
+    if (tableVersion > 8) {
+      assertEquals("timestamp-millis", tableSchema.getField("ts_millis").schema().getLogicalType().getName());
+    }
+    assertEquals("timestamp-micros", tableSchema.getField("ts_micros").schema().getLogicalType().getName());
+    if (tableVersion > 8 && !HoodieSparkUtils.isSpark3_3()) {
+      assertEquals("local-timestamp-millis", tableSchema.getField("local_ts_millis").schema().getLogicalType().getName());
+      assertEquals("local-timestamp-micros", tableSchema.getField("local_ts_micros").schema().getLogicalType().getName());
+    }
+
+    assertEquals("date", tableSchema.getField("event_date").schema().getLogicalType().getName());
+
+    if (tableVersion > 8) {
+      assertEquals("bytes", tableSchema.getField("dec_plain_large").schema().getType().getName());
+      assertEquals("decimal", tableSchema.getField("dec_plain_large").schema().getLogicalType().getName());
+      assertEquals(20, ((LogicalTypes.Decimal) tableSchema.getField("dec_plain_large").schema().getLogicalType()).getPrecision());
+      assertEquals(10, ((LogicalTypes.Decimal) tableSchema.getField("dec_plain_large").schema().getLogicalType()).getScale());
+    }
+    assertEquals("fixed", tableSchema.getField("dec_fixed_small").schema().getType().getName());
+    assertEquals(3, tableSchema.getField("dec_fixed_small").schema().getFixedSize());
+    assertEquals("decimal", tableSchema.getField("dec_fixed_small").schema().getLogicalType().getName());
+    assertEquals(5, ((LogicalTypes.Decimal) tableSchema.getField("dec_fixed_small").schema().getLogicalType()).getPrecision());
+    assertEquals(2, ((LogicalTypes.Decimal) tableSchema.getField("dec_fixed_small").schema().getLogicalType()).getScale());
+    assertEquals("fixed", tableSchema.getField("dec_fixed_large").schema().getType().getName());
+    assertEquals(8, tableSchema.getField("dec_fixed_large").schema().getFixedSize());
+    assertEquals("decimal", tableSchema.getField("dec_fixed_large").schema().getLogicalType().getName());
+    assertEquals(18, ((LogicalTypes.Decimal) tableSchema.getField("dec_fixed_large").schema().getLogicalType()).getPrecision());
+    assertEquals(9, ((LogicalTypes.Decimal) tableSchema.getField("dec_fixed_large").schema().getLogicalType()).getScale());
+
+    sqlContext.clearCache();
+    Dataset<Row> df = sqlContext.read()
+        .options(hudiOpts)
+        .format("org.apache.hudi")
+        .load(tableBasePath);
+
+    long totalCount = df.count();
+    long expectedHalf = totalCount / 2;
+    long tolerance = totalCount / 20;
+    if (totalCount < 100) {
+      tolerance = totalCount / 4;
+    }
```

Comment:
Should we also validate the column stats index entries directly against parquet min/max stats?